File: classification.f
C     ******************************************************************
C     ******************************************************************
C
C     This program filters the problems of CUTEr collection. For 
C     achieving this objective YOU MUST MODIFY it according to the type 
C     of problems you want to select. See below where your modifications
C     must be inserted. The program returns a text file called my_problems 
C     containing the names of filtered problems. 
C
C     See http://www.cuter.rl.ac.uk/Problems/classification.shtml for 
C     a complete explanation of the CUTE classification scheme.
C
C     For use this routine you must have a text file called CUTEr_problems
C     with the names of all problems (line by line), and other text file
C     called CUTEr_classification with the classifications of all problems
C     (line by line). You can get this information on the website 
C     http://www.cuter.rl.ac.uk/Problems/mastsif.shtml
C     A simple way is to copy all the table with name, files available
C     and CUTEr classification, paste into a xlsx file and extract the 
C     required information by copying and pasting the first and third
C     columns.
C
C     Author: Leandro Prudente
C     e-mail: lfprudente@gmail.com
C     Date:   12/03/2012 
C
C     If you find any error, please contact me by e-mail.
C

      program CUTEr_classification
      
      implicit none
      
      character(1) objective, constraint, smooth, origin, 
     +             eivariable
      character(8) problem, numvariable, numconstraint
      character(30) classification
      integer derivative, hifen, nvariable, nconstraint, nproblem,
     +        end_file
      logical fixnumvar, fixnumconst
      
C     Variables of the program:
C
C     The possible values of the variables follow the CUTE classification
C     scheme.
C     
C     objective    character, contains the type of the problem objective
C                  function. Possible values are:
C                  N no objective function is defined,
C                  C the objective function is constant,
C                  L the objective function is linear,
C                  Q the objective function is quadratic,
C                  S the objective function is a sum of squares, and
C                  O the objective function is none of the above. 
C
C     constraint   character, contains the type of constraints of the 
C                  problem. Possible values are:
C                  U the problem is unconstrained,
C                  X the problem's only constraints are fixed variables,
C                  B the problem's only constraints are bounds on the 
C                    variables,
C                  N the problem's constraints represent the adjacency 
C                    matrix of a (linear) network,
C                  L the problem's constraints are linear,
C                  Q the problem's constraints are quadratic, and
C                  O the problem's constraints are more general than any
C                    of the above alone.
C
C     smooth       character, indicates the smoothness of the problem. 
C                  There are two choices:
C                  R the problem is regular, that is its first and second
C                    derivatives exist and are continuous everywhere, or
C                  I the problem is irregular.
C
C     derivative,  integer, indicates the degree of the highest derivatives
C                  provided analytically within the problem description.
C                  It is restricted to be 0, 1 or 2.
C
C     origin,      character,  indicates the primary origin and/or interest
C                  of the problem. Possible values are:
C                  A the problem is academic, that is, has been constructed
C                    specifically by researchers to test one or more 
C                    algorithms,
C                  M the problem is part of a modelling exercise where the
C                    actual value of the solution is not used in a genuine
C                    practical application, and
C                  R the problem's solution is (or has been) actually used
C                    in a real application for purposes other than testing
C                    algorithms.
C
C     eivariable,  character, ndicates whether or not the problem 
C                  description contains explicit internal variables. 
C                  There are two possible values, namely
C                  Y the problem description contains explicit internal 
C                    variables, or
C                  N the problem description does not contain any explicit
C                    internal variables. 
C
C     nvariable,   integer, see below.
C
C     nconstraint, integer, see below.
C
C     fixnumvar,   logical, 
C                  true the number of variables in the problem is fixed
C                        and is available in nvariable,
C                  false the number of variables in the problem can be 
C                         chosen by the user.
C
C     fixnumconst, logical,
C                  true  the number of constraints in the problem is fixed
C                        and is available in nconstraint.
C                  false the number of constraints in the problem can be 
C                        chosen by the user
C
      
      open(10,file='CUTEr_problems')
      open(20,file='CUTEr_classification')
      open(30,file='my_problems')
      
      nproblem = 0

 100  continue
      
          read(10,'(A8)', iostat = end_file) problem
          
          if ( end_file .ne. 0 ) stop
          
          nproblem = nproblem + 1
          
          read(20,'(A30)')classification
          
          objective  = classification(1:1)
          constraint = classification(2:2)
          smooth     = classification(3:3)
          read(classification(4:4),*) derivative
          origin     = classification(6:6)
          eivariable = classification(7:7)
          
          hifen = scan(classification,'-',back = .true.)
          
          numvariable   = classification(9:hifen-1)
          numconstraint = classification(hifen+1:30)
          
          if ( scan(numvariable,'V') .eq. 0 ) then 
              fixnumvar = .true.
              read(numvariable,*) nvariable
          else
              fixnumvar = .false.
          end if
          
          if ( scan(numconstraint,'V') .eq. 0 ) then 
              fixnumconst = .true.
              read(numconstraint,*) nconstraint
          else
              fixnumconst = .false.
          end if
          
C     ******************************************************************
C     FROM HERE ON YOU MUST MODIFY THE PROGRAM TO SELECT THE PROBLEMS 
C     ACCORDING TO YOUR DESIRE.
C     ******************************************************************

C     In this exemple we will select the problems with linear or quadratic
C     objective function and constraints and with at least ten variables 
C     (or the problems for wich we can choose the number of variables).

      if ( ( scan(objective,'LQ') .ne. 0 ) .and.  
     +     ( scan(constraint,'LQ') .ne. 0 ) ) then
          if ( .not. fixnumvar ) then
              write(30,'(A8)') problem
          else if ( nvariable .ge. 10 ) then
              write(30,'(A8)') problem
          end if            
      end if
          
C     ******************************************************************
C     STOP HERE YOUR MODIFICATIONS.
C     ******************************************************************          
      
      go to 100
    
      end program

File: classification.py
"""
Copyright (C) 2012 Raniere Silva <ra092767@ime.unicamp.br>

This program is free software: you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation, either version 3 of the License, or (at your
option) any later version.

This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
General Public License for more details.

You should have received a copy of the GNU General Public License along
with this program. If not, see <http://www.gnu.org/licenses/>.
"""

import sys

def get_info():
    """
    Get Information from the site.
    """
    from subprocess import call
    call("./get_files.sh", shell=True)

def _w2f():
    """
    Write To File

    This function wait the user input the name of the wanted file for output.
    """
    print("Choose the name of the file where the output will be write.")
    tmp = input("Default is stdout:")
    return tmp
    
def _tof():
    """
    Type of problem Objective Function

    This function wait the uset input the wanted type of objective
    function.
    """
    print("Choose the type of the problem objective function.\n")
    print("\tN - no objective function is defined,")
    print("\tC - the objective function is constant,")
    print("\tL - the objective function is linear,")
    print("\tQ - the objective function is quadratic,")
    print("\tS - the objective function is a sum of squares,")
    print("\tO - the objective function is none of the above.\n")
    tmp = input("Default 'NCLQSO':")
    if not tmp:
        tmp = 'NCLQSO'
    return tmp

def _tc():
    """
    Type of Constraints of the problem

    This function wait the uset input the wanted type of constraints of the
    problem.
    """
    print("Choose the type of constraints of the problem.\n")
    print("\tU - the problem is unconstrained,")
    print("\tX - the problem's only constraints are fixed variables,")
    print("\tB - the problem's only constraints are bounds on the variables,")
    print("\tN - the problem's constraints represent the adjacency matrix of a (linear) network,")
    print("\tL - the problem's constraints are linear,")
    print("\tQ - the problem's constraints are quadratic,")
    print("\tO - the problem's constraints are more general than any of the above alone.\n")
    tmp = input("Default 'UXBNLQO':")
    if not tmp:
        tmp = 'NCLQSO'
    return tmp

def _ts():
    """
    Type of the Smoothness of the problem.

    This function wait the user input the wanted type of smoothness.
    """
    print("Choose the smoothness of the problem.\n")
    print("\tR - the problem is regular, that is its first and second "
    "derivatives exist and are continuous everywhere, or")
    print("\tI - the problem is irregular.\n")
    tmp = input("Default 'RI':")
    if not tmp:
        tmp = 'RI'
    return tmp

def _dhd():
    """
    Degree of the Highest Derivatives

    This function wait the user input the degree of the highest
    derivatives provided analytically.
    """
    print("Choose the degree of the highest derivatives provided "
    "analytically within the problem description. It is restricted to being "
    "0, 1 or 2.\n")
    tmp = input("Default '012':")
    if not tmp:
        tmp = '012'
    return tmp

def _torigin():
    """
    Type of Origin of the problem

    This function wait the user input the wanted type of origin of the problem.
    """
    print("Choose the primary origin and/or interest of the problem.\n")
    print("\tA - the problem is academic, that is, has been constructed "
    "specifically by researchers to test one or more algorithms,")
    print("\tM - the problem is part of a modelling exercise where the actual "
    "value of the solution is not used in a genuine practical "
    "application,")
    print("\tR the problem's solution is (or has been) actually used in a real "
    "application for purposes other than testing algorithms.\n")
    tmp = input("Default 'AMR':")
    if not tmp:
        tmp = 'ARM'
    return tmp

def _eiv():
    """
    Explicit Internal Variable

    This function wait the user input the wanted the configuration for the
    explicit internal variables.
    """
    print("Choose if the problem description contains explicit internal "
    "variables or not.\n")
    print("\tY - the problem description contains explicit internal "
    "variables,")
    print("\tN - the problem description does not contain any explicit internal "
    "variables.")
    tmp = input("Default 'YN':")
    if not tmp:
        tmp = 'YN'
    return tmp
 
def interact():
    print("Py-CUTEr-Problem-Chooser Copyright (C) 2012 Raniere Silva")
    print("This program comes with ABSOLUTELY NO WARRANTY.")
    print("This is free software, and you are welcome to redistribute it")
    print("under certain conditions; type `show c' for details.\n")

    w2f = _w2f()
    tof = _tof()
    tc = _tc()
    ts = _ts()
    dhd = _dhd()
    torigin = _torigin()
    eiv = _eiv()

    classification(tof, tc, ts, dhd, torigin, eiv)

def classification(tof='NCLQSO', tc='UXBNLQO', ts='RI', dhd='012',
        torigin='ARM', eiv='YN', w2f=''):
    # get_info()
    f_name = open('CUTEr_problems')
    f_clas = open('CUTEr_classification')
    if w2f:
        f_out = open(w2f, 'w')

    for p in f_name:
        choosed = True
        p_clas = f_clas.readline()
        p_clas = p_clas.split('-')
        if p_clas[0][0] not in tof:
            choosed = False
        elif p_clas[0][1] not in tc:
            choosed = False
        elif p_clas[0][2] not in ts:
            choosed = False
        elif p_clas[0][3] not in dhd:
            choosed = False
        elif p_clas[1][0] not in torigin:
            choosed = False
        elif p_clas[1][1] not in eiv:
            choosed = False
        if choosed:
            if w2f:
                f_out.write('{0}\n'.format(p.rstrip()))
            else:
                print('{0}'.format(p.rstrip()))

    f_name.close()
    f_clas.close()
    if w2f:
        f_out.close()

if __name__ == "__main__":
    """Call the classification from the command line interpreter. ::

        $ python classification.py --help
    """
    import argparse
    from argparse import RawTextHelpFormatter

    if sys.version_info[0] < 3:
        raise Exception('It require Python 3.0 or later.')

    # Parse of flags.
    parser = argparse.ArgumentParser(description='Choose problems from CUTer.',
            formatter_class=RawTextHelpFormatter)
    parser.add_argument('-i', action='store_true',
            help='start interact mode.')
    parser.add_argument('-f', type=str, default='',
            help='Specify the file where the output will be write.')
    parser.add_argument('-tof', type=str, default='NCLQSO',
            help="Select type of objective function. Default is 'NCLQSO'.\n"
            "N - no objective function is defined,\n"
            "C - the objective function is constant,\n"
            "L - the objective function is linear,\n"
            "Q - the objective function is quadratic,\n"
            "S - the objective function is a sum of squares,\n"
            "O - the objective function is none of the above.\n")
    parser.add_argument('-tc', type=str, default='UXBLQO',
            help="Select the type of constraints of the problem. Default is 'UXBLQO'.\n"
            "U - the problem is unconstrained,\n"
            "X - the problem's only constraints are fixed variables,\n"
            "B - the problem's only constraints are bounds on the variables,\n"
            "N - the problem's constraints represent the adjacency matrix of a (linear) network,\n"
            "L - the problem's constraints are linear,\n"
            "Q - the problem's constraints are quadratic,\n"
            "O - the problem's constraints are more general than any of the above alone.\n")
    parser.add_argument('-ts', type=str, default='RI',
            help="Select the smoothness of the problem. Default is 'RI'\n"
            "R - the problem is regular, that is its first and second "
            "derivatives exist and are continuous everywhere, or\n"
            "I - the problem is irregular.\n")
    parser.add_argument('-dhd', type=str, default='012',
            help="Select the degree of the highest derivatives provided "
    "analytically within the problem description. It is restricted to being "
    "0, 1 or 2.\n")
    parser.add_argument('-to', type=str, default='AMR',
            help="Select the primary origin and/or interest of the problem.\n"
            "A - the problem is academic, that is, has been constructed "
            "specifically by researchers to test one or more algorithms,\n"
            "M - the problem is part of a modelling exercise where the actual "
            "value of the solution is not used in a genuine practical "
            "application,\n"
            "R the problem's solution is (or has been) actually used in a real "
            "application for purposes other than testing algorithms.\n")
    parser.add_argument('-eiv', type=str, default='YN',
            help="Select if the problem description contains explicit internal "
            "variables or not.\n"
            "Y - the problem description contains explicit internal "
            "variables,\n"
            "N - the problem description does not contain any explicit internal "
            "variables.")

    args = parser.parse_args()

    if args.i:
        interact()
    else:
        classification(args.tof, args.tc, args.ts, args.dhd, args.to,
                args.eiv, args.f)
